### ISSUE TITLE
Type intersection is not detected as empty through generics

### ISSUE THREAD

**Issue by fefrei** [microsoft/TypeScript#36736]
**TypeScript Version:**  3.7.5

**Search Terms:** union intersection type conflict empty inference generics

**Expected behavior:**

TypeScript can determine when type intersections are empty. For example,

```ts
const t1 = (t: "a" | ("b" & "c")): "a" => t;
const t2 = (t: number | ("b" & "c")): number => t;
```
is perfectly valid. This is very useful, and it even works through generics:

```ts
type Container<Type extends string> = {
    type: Type;
}

const t3 = (t: Container<"a"> | (Container<"b"> & { dataB: boolean } & Container<"a">)): Container<"a"> => t;
```

I'd expect `t3` to be validly typed even when the left side of the union and the return type are changed to another type, like so:

```ts
const t4 = (t: number | (Container<"b"> & { dataB: boolean } & Container<"a">)): number => t;
```

**Actual behavior:**

However, this is a type error:

```plain
Type 'number | (Container<"b"> & { dataB: boolean; } & Container<"a">)' is not assignable to type 'number'.
```

You can find context for where this is relevant [here on Stack Overflow](https://stackoverflow.com/questions/60168067/typescript-select-element-of-an-array-by-type-property), or in [this playground](https://www.typescriptlang.org/play/?ssl=28&ssc=1&pln=35&pc=1#code/C4TwDgpgBAwg9gO2AQwJYIgJwDwBVzQQAewECAJgM5SXCboDmAfFALxQDeAUFL1KJABcUfJADcXAL5cuA6AEF4SNBkxtYiFOizYARMl0sAZJx59yyFPOEIArgFsARlgnTZBKACElW1ep8qOrqOhlAm3HxQFiiewo5wcAA2EMgIrjIAxoi0UKiUAdpq7HgexKQU1LT0CMwAFBnCBarYVYxMADT8BMKiEACUwhm51E06vSysLBF8mBDAtpgIXZBwAGZQQ6xbULpwjgBWEBnAumEmGQB0cmzbcq4SXFkIOdHIwrWKmoFqAD5eo5g+gBtAC66iBZl4HGWEGE+l0nVe1igAEYAOxQSTtSGcGFwkKIyzIWL8TC2aDSEEPJ45OTkADic08IF66mwANZZTIVA0ykKLTobU6MAAFnBKGQoFyKrzfFhOSRuSMvvz0KssCIWAB+ERQGwQABuWCYtVe7w5HhMAIF1WYwJBnTkjTFEoQA1l31ZVpVzVF4rIEymONWcDU9WywA2UDWUSJfVMkUiqHWtTyAPqjoIfXjs3miw2Eki0kiwBFmDgAHcoBgqwBRTDlzC1AAGADk4FLkvYyJGY9cACQcOSSKAh2wUC7Nvr3TIR2NWYSfPl+dh0xnAZm9U1Ezrw6dcIA).

**Related Issues:** None found.

**Code**
```ts
const t1 = (t: "a" | ("b" & "c")): "a" => t;

type Container<Type extends string> = {
    type: Type;
}

const t2 = (t: Container<"a"> | (Container<"b"> & Container<"c">)): Container<"a"> => t;
const t3 = (t: Container<"a"> | (Container<"b"> & { dataB: boolean } & Container<"a">)): Container<"a"> => t;
const t4 = (t: number | (Container<"b"> & { dataB: boolean } & Container<"a">)): number => t;
```

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/#code/MYewdgzgLgBFCMMC8MAUUBcMBEBDbMAPmtgEYEBkOw2AlLVngUgHxwDcAUJ1AJ4AOAUxgBhcFFwBLMIIBOAHgAqA4YIAeUQWAAmEGNFnSA5mxQBvTjCtwVWZUK4BfbqEiwoAJmRpMo8VJkFJjZiVDEwCWk5eTJsNipwyMCYmhZ6LESA6ODkNiguV2g4AGZvdAz-KKD8ELRMqpjyeJgzGG1cCQAhLFIQEAAbQVwwGEcYBMrk4PS-CKzquNyOTkL3ABYy3zAAVwBbUjkiOsnspvGWto7cbphegaGRsYm5humGGB39w9ZloA)
